### PR TITLE
[bench] fio engine for objectstore

### PR DESCRIPTION
**Pull request by cbodley**
This pull request introduces an external fio engine for running workloads directly against the ObjectStore interface. Unlike other fio engines, it builds inside the ceph tree and produces a shared library. To run it with fio, you set engine=path/to/libfio_ceph_objectstore.so. An example fio job file is provided in src/test/objectstore.fio.

### CODE REVIEW

File: src/test/fio_ceph_objectstore.cc [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+  }
+};
+
+struct OnApplied : public Context {
+  struct io_u *io_u;
+  ObjectStore::Transaction *t;
+  OnApplied(struct io_u* io_u, ObjectStore::Transaction *t) : io_u(io_u), t(t) {}
+  void finish(int r) {
+
+	struct fio_ceph_os_iou *fio_ceph_os_iou =
```
Comment (xiaoxichen) [inline src/test/fio_ceph_objectstore.cc]: The indent looks doesn't right.

Could you pls fix it follow 

// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
// vim: ts=8 sw=2 smarttab

File: src/test/fio_ceph_objectstore.cc [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+	}
+
+  ObjectStore *fs = ObjectStore::create(g_ceph_context,
+			o->objectstore, td->o.directory,
+			o->filestore_journal ? o->filestore_journal : "");
+	if (fs == NULL) {
+		cout << "bad objectstore type " << o->objectstore << std::endl;
+		return 1;
+	}
+	if (fs->mkfs() < 0) {
```
Comment (xiaoxichen) [inline src/test/fio_ceph_objectstore.cc]: objectstore->mkfs() doesn't cleanup the previous old data, mostly the k/v database

**Comment by xiaoxichen** [comment]
I had an error on fio-objectstore + newstore, after debugging the code, I found the issue is when the second time I run the FIO, the mkfs() will not cleanup the data in previous run(which seems by design also in FlieStore/Keyvaluestore).  Then in your code you are tying to create the collection(pg) agian, thus causing an assert.


2015-09-28 10:13:40.863357 7facbcd93840  1 newstore(/mnt/) mkfs path /mnt/
2015-09-28 10:13:40.863373 7facbcd93840  1 newstore(/mnt/) _open_path using fs driver 'generic'
2015-09-28 10:13:40.863408 7facbcd93840  1 newstore(/mnt/) mkfs fsid is already set to b09bebab-6148-41ae-95d3-a9d7ba8db4b2
2015-09-28 10:13:40.869700 7facbcd93840  1 newstore(/mnt/) _open_db opened rocksdb path /mnt/ options 
2015-09-28 10:13:40.869709 7facbcd93840 10 newstore(/mnt/) mkfs success
2015-09-28 10:13:40.870107 7facbcd93840  1 newstore(/mnt/) mount path /mnt/
2015-09-28 10:13:40.870121 7facbcd93840  1 newstore(/mnt/) _open_path using fs driver 'generic'
2015-09-28 10:13:40.874838 7facbcd93840  1 newstore(/mnt/) _open_db opened rocksdb path /mnt/ options 
2015-09-28 10:13:40.875134 7facbcd93840  1 newstore(/mnt/) _recover_next_fid old fid_max 1/7168
2015-09-28 10:13:40.875175 7facbcd93840  1 newstore(/mnt/) _recover_next_nid old nid_max 7168
2015-09-28 10:13:40.875265 7facbcd93840 20 newstore(/mnt/) _open_collections opened 0.0_head
2015-09-28 10:13:40.875287 7facbcd93840 10 newstore(/mnt/) _aio_start
2015-09-28 10:13:40.875432 7facbcd93840 10 newstore(/mnt/) _wal_replay start
2015-09-28 10:13:40.875455 7facbcd93840 10 newstore(/mnt/) _wal_replay completed 0 events
2015-09-28 10:13:40.875479 7facb4a9b700 10 newstore(/mnt/) _aio_thread start
2015-09-28 10:13:40.876602 7fac90ee7700 10 newstore(/mnt/) _kv_sync_thread start
2015-09-28 10:13:40.876713 7fac90ee7700 20 newstore(/mnt/) _kv_sync_thread sleep
2015-09-28 10:13:40.876731 7facbcd93840  5 newstore(/mnt/) queue_transactions new osr(default 0x19e40b8)/0x19e40b8
2015-09-28 10:13:40.876746 7facbcd93840 20 newstore(/mnt/) _txc_create osr 0x1a1cab0 = 0x1a1cce0

**Comment by xiaoxichen** [comment]
os/newstore/NewStore.cc: In function 'int NewStore::_txc_add_transaction(NewStore::TransContext*, ObjectStore::Transaction*)' thread 7facbcd93840 time 2015-09-28 10:13:40.876874
os/newstore/NewStore.cc: 2978: FAILED assert(!c)
 ceph version 9.0.3-1576-gf8797b7 (f8797b7d7252a4b26ddc72cef1878a1527ec294b)
 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x8b) [0x7faca31661eb]
 2: (NewStore::_txc_add_transaction(NewStore::TransContext*, ObjectStore::Transaction*)+0x265b) [0x7faca2d1cddb]
 3: (NewStore::queue_transactions(ObjectStore::Sequencer*, std::list<ObjectStore::Transaction*, std::allocator<ObjectStore::Transaction*> >&, std::shared_ptr<TrackedOp>, ThreadPool::TPHandle*)+0x33d) [0x7faca2d1d53d]
 4: (ObjectStore::apply_transactions(ObjectStore::Sequencer*, std::list<ObjectStore::Transaction*, std::allocator<ObjectStore::Transaction*> >&, Context*)+0x1c5) [0x7faca2e0ee15]
 5: (()+0x2b9ff8) [0x7faca2ce6ff8]
 6: (td_io_init()+0x1b) [0x40e0bb]
 7: ../../fio/fio/fio() [0x44d2db]
 8: ../../fio/fio/fio() [0x44f98c]
 9: (fio_backend()+0xfb) [0x44fe7b]
 10: (main()+0x70) [0x40c1c0]
 11: (__libc_start_main()+0xf5) [0x7facb6e65ec5]
 12: ../../fio/fio/fio() [0x40c20c]
 NOTE: a copy of the executable, or `objdump -rdS <executable>` is needed to interpret this.
2015-09-28 10:13:40.878598 7facbcd93840 -1 os/newstore/NewStore.cc: In function 'int NewStore::_txc_add_transaction(NewStore::TransContext*, ObjectStore::Transaction*)' thread 7facbcd93840 time 2015-09-28 10:13:40.876874
os/newstore/NewStore.cc: 2978: FAILED assert(!c)

**Comment by yuyuyu101** [comment]
do we consider to make fio as submodule? So it may let compile with fio source dir ease

File: src/test/fio_ceph_objectstore.cc (hunk 1 above)
Comment (cbodley) [inline src/test/fio_ceph_objectstore.cc]: I added the mode line and reindented the file, thanks

File: src/test/fio_ceph_objectstore.cc (hunk 2 above)
Comment (cbodley) [inline src/test/fio_ceph_objectstore.cc]: I implemented fio's close_file() hook by removing the associated objects, so it should clean up after itself now.

**Comment by cbodley** [comment]
@xiaoxichen 
> Then in your code you are tying to create the collection(pg) agian, thus causing an assert.

It will no longer call create_collection() if the collection already exists.

**Comment by cbodley** [comment]
@yuyuyu101 
> do we consider to make fio as submodule?
Because fio support is disabled by default in the build systems, I don't think that most developers will be interested in building it. And in my experience, fio has been really easy to clone and build. So I would personally prefer not to make it a submodule.

**Comment by xiaoxichen** [comment]
Do we have any chance that directly read the ceph.conf instead of configuration every ceph parameters in the code?A bit complicated but should be better as the objectstore has a lot of tunables..

one more question, now we only have one collection(PG), so even we have multithread configured in fio, they all work in the same PG, which is not the ceph's syle. is that practicable to have pre-configurable PG count, and threads randomly pickup a PG and work on it?

**Comment by cbodley** [comment]
@xiaoxichen, we're calling global_init(), so it will attempt to load a config file from the default locations. You can also use the environment variable CEPH_CONF= to override the path.

**Comment by yuyuyu101** [comment]
I guess the better idea may add ceph conf path as fio option, so we can pass ceph conf path to global_init. Then we could discard other ceph options.

**Comment by cbodley** [comment]
> one more question, now we only have one collection(PG), so even we have multithread configured in fio, they all work in the same PG, which is not the ceph's syle. is that practicable to have pre-configurable PG count, and threads randomly pickup a PG and work on it?

@xiaoxichen Sorry I missed this part earlier, I agree that we need a way to run against multiple collections/sequencers. I like your suggestion to spread the objects within a job across a number of collections up to osd_pool_default_pg_num.

I'm also looking into how we support multiple simultaneous jobs. We need to make sure there's only one instance of the ObjectStore at any time, which requires changes to a) share the ObjectStore instance between jobs within a single fio process, and b) prevent fio from forking additional processes (which it does by default).

**Comment by ktdreyer** [comment]
> do we consider to make fio as submodule?

On the contrary, this seems like a good feature to submit to fio upstream

**Comment by cbodley** [comment]
@ktdreyer We could certainly go in that direction, but I see two obstacles at the moment.

First, that the ObjectStore isn't a public interface. There was an effort to make it pluggable, which I assume would involve adding a public header, but I haven't seen any progress on that lately.

Second is that fio's build system doesn't support c++, though I do have patches (https://github.com/cohortfsllc/fio/commit/fc42e51b2e2d03c19c24e8a35c9e4849d502c715 and https://github.com/cohortfsllc/fio/commit/1882dcbfde481f5b142f1fbfdb40e29153aa1323) that we could push upstream.

**Comment by mattbenjamin** [comment]
The lack of a C++ ABI would seem to make upstreaming a C++ build info fio problematic.  I don't see imposing a C interface on things as a solution, as it runs counter to the intended purpose of this integration, which wants to minimize or eliminate the thickness of the binding.  I guess a submodule would allow arbitrary adaptation, but we probably wouldn't want to support actually adapting it.  So, in fine, I think accepting the PR approach brings the most benefit (and avoids distracting upstreaming and adaptation support work).

**Comment by cbodley** [comment]
There are currently some issues with the build that I'm looking into:

* the automake build fails to link against libos due to lack of PIC,
* upstream fio recently introduced a couple things to their headers (typedef int bool; and void *private;) that don't compile as c++

**Comment by liewegas** [comment]
close this for now

**Pull request closed without merging**